src/audio/OpenALAudioEngine.cpp

```diff
@@ -0,0 +1,485 @@
+         * @param buf The buffer to render data into
+         * @return true The source has more data to render
+         * @return false The source has no more data to render
+         */
+        bool renderBuffer(std::array<std::int16_t, BUFFER_SIZE>& buf)
+        {
+            int rendered = m_stream(buf.data(), BUFFER_SIZE);
+            if(rendered < 0) return false;
+
+            while(rendered < BUFFER_SIZE)
+            {
+                if(m_stop) return false;
+                int res = m_stream(buf.data() + rendered, BUFFER_SIZE - rendered);
+                if(res < 0) return false;
+                rendered += res;
+            }
+            return true;
+        }
+
+        void renderThread()
+        {
+            std::array<std::int16_t, BUFFER_SIZE> dataBuffer{};
+
+            for(auto buf : m_buffers)
+            {
+                if(!renderBuffer(dataBuffer)) return;
+                alBufferData(buf, m_format, dataBuffer.data(), BUFFER_SIZE * 2, m_sampleRate);
+            }
+            
+            alSourceQueueBuffers(m_source, BUFFER_NUM, m_buffers.data());
+
+            if(!renderBuffer(dataBuffer)) return;
+
+            while(true)
+            {
+                if(m_stop) return;
+                if(!m_hasSource) continue;
+
+                ALint processedBuffers;
+                alGetSourcei(m_source, AL_BUFFERS_PROCESSED, &processedBuffers);
+                if(processedBuffers <= 0) continue;
+                while(processedBuffers--)
+                {
+                    ALuint buffer;
+                    alSourceUnqueueBuffers(m_source, 1, &buffer);
+                    alBufferData(buffer, m_format, dataBuffer.data(), BUFFER_SIZE * 2, m_sampleRate);
+                    alSourceQueueBuffers(m_source, 1, &buffer);
+                    if(!renderBuffer(dataBuffer)) return;
+                }
+
+                if(state() != State::Playing) play();
+            }
+        }
+
+        void getSource() override
+        {
+            if (m_hasSource) return;
+            m_source = m_engine->getFreeSource();
+            m_hasSource = true;
+        }
+    };
+}  // namespace Audio
+
+SoundPtr OpenALAudioEngine::createSound(const std::int16_t* buf, std::size_t len, Format fmt, std::size_t sampleRate)
+{
+    captureContext();
+    auto ptr = std::make_shared<OpenALSound>(this, buf, len, fmt, sampleRate);
+    m_bufferedSounds.push_back(ptr);
+    return ptr;
+}
+
+SoundPtr OpenALAudioEngine::createSound(SoundStream stream, Format fmt, std::size_t sampleRate)
+{
+    captureContext();
+    auto ptr = std::make_shared<StreamingOpenALSound>(this, stream, fmt, sampleRate);
+    m_streamingSounds.push_back(ptr);
+    return ptr;
+}
+
+ALuint OpenALAudioEngine::getFreeSource()
+{
+    captureContext();
+
+    auto lambda = [](const auto& sound) { return sound.use_count() == 1; };
+
+    // Remove all expired sounds
+    std::remove_if(std::begin(m_bufferedSounds), std::end(m_bufferedSounds), lambda);
```

Comment:
Btw Is this needed? Actually it just moves objects in vector.

Reply:
What do you mean if it is needed? The `shared_ptr`s are never going to get deleted if they stay inside the array, this allows them to be freed

Reply:
I think about missing `clear`, https://cpppatterns.com/patterns/remove-elements-from-container.html